compiler/modules/nback/structure.bal

```diff
@@ -352,30 +352,48 @@ function mappingOrderFields(t:MappingAtomicType mat, string[] fieldNames, bir:Op
+        RuntimeFunction rf = fill ? mappingFillingGetFunction : mappingGetFunction;
+        llvm:Value memberWithErr = <llvm:Value>builder.call(scaffold.getRuntimeFunctionDecl(rf), [mapping, check buildString(builder, scaffold, keyOperand)]);
+        llvm:Value member = fill ? buildCheckPanicCode(builder, scaffold, memberWithErr, insn.pos) : memberWithErr;
+        buildStoreMappingGet(builder, scaffold, insn, member, mapping, mappingReg, keyOperand);
+    }
+    else if fieldIndexRes.isAtomicOpen {
+        llvm:BasicBlock bbExact = scaffold.addBasicBlock();
+        llvm:BasicBlock bbInexact = scaffold.addBasicBlock();
+        llvm:BasicBlock bbJoin = scaffold.addBasicBlock();
+        llvm:Value isExact = buildIsExact(builder, scaffold, mapping);
+        builder.condBr(isExact, bbExact, bbInexact);
+        builder.positionAtEnd(bbExact);
+        buildNoPanicStoreMappingGet(builder, scaffold, insn, mapping, mappingReg, keyOperand, mappingIndexedGetFunction, llvm:constInt(LLVM_INT, fieldIndexRes.fieldIndex));
+        builder.br(bbJoin);
+        builder.positionAtEnd(bbInexact);
+        buildNoPanicStoreMappingGet(builder, scaffold, insn, mapping, mappingReg, keyOperand, mappingGetFunction, check buildString(builder, scaffold, keyOperand));
+        builder.br(bbJoin);
+        builder.positionAtEnd(bbJoin);
     }
     else {
-        fill = false;
-        rf = mappingIndexedGetFunction;
-        k = llvm:constInt(LLVM_INT, fieldIndex);
+        buildNoPanicStoreMappingGet(builder, scaffold, insn, mapping, mappingReg, keyOperand, mappingIndexedGetFunction, llvm:constInt(LLVM_INT, fieldIndexRes.fieldIndex));
     }
-    llvm:Value mapping = builder.load(scaffold.address(mappingReg));
-    llvm:Value memberWithErr = <llvm:Value>builder.call(scaffold.getRuntimeFunctionDecl(rf), [mapping, k]);
-    llvm:Value member = fill ? buildCheckPanicCode(builder, scaffold, memberWithErr, insn.pos) : memberWithErr;
+}
+
+function buildNoPanicStoreMappingGet(llvm:Builder builder, Scaffold scaffold, bir:MappingGetInsn insn, llvm:Value mapping, bir:Register reg, bir:StringOperand keyOperand, RuntimeFunction rf, llvm:Value k) {
+    llvm:Value member = <llvm:Value>builder.call(scaffold.getRuntimeFunctionDecl(rf), [mapping, k]);
+    buildStoreMappingGet(builder, scaffold, insn, member, mapping, reg, keyOperand);
+}
+
+function buildStoreMappingGet(llvm:Builder builder, Scaffold scaffold, bir:MappingGetInsn insn, llvm:Value mappingMember, llvm:Value mapping, bir:Register reg, bir:StringOperand keyOperand) {
     t:SemType resultType = insn.result.semType;
+    llvm:Value member = mappingMember;
     if isPotentiallyExact(resultType) {
-        if !isMappingMemberTypeExact(scaffold.typeContext(), mappingReg.semType, keyOperand, resultType) {
-            member = buildClearExact(builder, scaffold, member, resultType);
+        if !isMappingMemberTypeExact(scaffold.typeContext(), reg.semType, keyOperand, resultType) {
```

Comment:
Shall we keep the original name `mappingReg`?

Reply:
Changed